### PR TITLE
staking: avoid proportional slashing leak dust into chunks that should not be slashed

### LINKED ISSUE #11810: Rewrite proportional slashing to to not leak dust into chunks that should not be slashed.
The purpose of proportional slashing, as per #10982 and https://github.com/paritytech/substrate/pull/11782 to slash certain chunks and the active ledger, and leave the rest untouched. 

This helps applications like nomination pools on top of the current staking provide better guarantees to their users.

For example, in `Ledger { active: 10, unlokcing: vec![(10, 10) (10, 11)] }`, we want to slash 5 from `active` and 5 from first unlocking chunk, instead of slashing 10 out of `active`. 

The current implementation does achieve this, but since all slashes are rounded down, it is very likely that the sum of the proportional slashes are not enough to cover the slash, and some very small dust amount is slashed from the "first chunk that should have not been affected".

Examples of this can be found in `proportional_ledger_slash_works` tests.

This is not a big deal, but fixing it will clean up both the code and the interface a bit. All changes must be contained in `fn slash` in `impl Ledger`.

### PR DESCRIPTION
**Pull request by NingLin-P**
Attempt to fix #11810 

This PR replaces the slash ratio with the remaining ratio, since calculations are rounded down, the `slash_from_target` calculated from the remaining ratio will be a bit more than the expected proportion instead of less, thus fixing the issue. This PR also comes with a little refactoring.

polkadot address: 1CWyso9Zw46QdR54sTvKJXTUdmZznYZa2aJfNHdG6xQ6L71

### CODE REVIEW

**Comment by NingLin-P** [comment]
@kianenigma PTAL

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1752357

File: frame/staking/src/lib.rs [hunk 1]
```diff
@@ -570,36 +570,46 @@ impl<T: Config> StakingLedger<T> {
...
-				)
-			} else {
-				// We just slash from the last chunk to the most recent one, if need be.
-				(None, (0..self.unlocking.len()).rev().collect::<Vec<_>>())
+		// if `remaining_ratio` is not zero this is a proportional slash, each chunk and
+		// `self.active` will be slashed proportionally based on the `remaining_ratio`, slashed to
+		// zero otherwise. In both cases, we slash first `self.active`, and then
+		// `slash_chunks_priority`.
+		let (remaining_ratio, slash_chunks_priority) = {
+			let mut affected_balance = BalanceOf::<T>::zero();
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:578]: I don't quite agree with the changes here, like `let mut first_chunk = None;` and assigning to it in the loop. This is actually not Rust-idiomatic at all, and the previous `if let Some(_) = iterator {} else {}` is more preferred.

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -611,21 +621,19 @@ impl<T: Config> StakingLedger<T> {
 			slash_era,
 			self,
 			slash_chunks_priority,
-			maybe_proportional,
+			Perquintill::one() - remaining_ratio,
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:624]: This should be safe math. Or, yet again, I want to complain that the refactor is more of step backwards, since `Option` is the prefect type to show a `maybe_it_is_proportional`, instead of some value being `0`.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Right direction and we can fix this, but please: 

1. The refactors are a bit opinionated, let's keep this PR limited only to fixing the rounding down (unless if you want to try a different approach at refactoring the code)
2. As a simpler approach, please use the same logic as before, but use `Rounding::Up` when multiplying the Prequintill. 

The combination of these two should make the diff of this PR very small, which makes it way easier to merge, since it is a very sensitive code. Thanks!

File: frame/staking/src/lib.rs (hunk 1 above)
Comment (NingLin-P) [reply frame/staking/src/lib.rs:578]: I thought using a single loop to get `first_index` and `affected_balance` was more efficient (less iteration) and cleaner (less `if else` branch), but indeed hand code the functionality that the `Iterator` API already provided is not Rust-idiomatic.

File: frame/staking/src/lib.rs (hunk 2 above)
Comment (NingLin-P) [reply frame/staking/src/lib.rs:624]: The intention was to unify both cases (proportional or not) under a  ratio,  proportional or not just a matter of different ratio, but since it is important to get clear of these cases this indeed may not be a good idea

**Comment by NingLin-P** [comment]
Thanks for the hint @kianenigma !

> As a simpler approach, please use the same logic as before, but use Rounding::Up when multiplying the Prequintill.

I found there may have two places need to change to `Rounding::Up`:
https://github.com/paritytech/substrate/blob/e7aa858767c2f19a1f084b98f7a43db52b53b969/frame/staking/src/lib.rs#L595
https://github.com/paritytech/substrate/blob/e7aa858767c2f19a1f084b98f7a43db52b53b969/frame/staking/src/lib.rs#L619

the first place may have more impact to the result, it can changing to `Perquintill::from_rational_with_rounding` but which return a `Result` I'm not quite sure the proper way to handle the error here. The second place can easily changing to `ratio.mul_ceil(*target)` though.

**Comment by kianenigma** [comment]
You should do what normal `from_rational` is doing: https://github.com/paritytech/substrate/blob/0ebeac464ecacfb1b9cb36b63ed0040339807cd3/primitives/arithmetic/src/per_things.rs#L332

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1757628

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1757627

File: frame/nomination-pools/test-staking/src/lib.rs [hunk 3]
```diff
@@ -490,12 +490,10 @@ fn pool_slash_proportional() {
 		assert_eq!(
 			pool_events_since_last_call(),
 			vec![
-				// This last pool got slashed only the leftover dust. Otherwise in principle, this
-				// chunk/pool should have not been affected.
-				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 127, balance: 19 },
-				// This pool got slashed 12.5, which rounded down to 12.
-				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 128, balance: 8 },
-				// This pool got slashed 12.5, which rounded down to 12.
+				// This pool got slashed 12.5, which rounded down to 13.
```
Comment (kianenigma) [inline frame/nomination-pools/test-staking/src/lib.rs:493]:
```suggestion
				// This pool got slashed 12.5, which rounded up to 13.
```

File: frame/staking/src/tests.rs [hunk 4]
```diff
@@ -5233,26 +5233,12 @@ fn proportional_ledger_slash_works() {
 	ledger.active = unit;
 	ledger.total = unit * 4 + value;
 	// When
-	assert_eq!(ledger.slash(slash, 0, 0), slash - 43);
+	assert_eq!(ledger.slash(slash, 0, 2), slash);
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5236]: Please add this test case and make sure it is reasonable as well: https://github.com/paritytech/substrate/commit/bdd139989cb4dfa8cc6cc6b5ee470a711df32d25

Ideally, existing tests should not be changed at all.

Comment (kianenigma) [reply frame/staking/src/tests.rs:5236]: Not sure why you changed this test?

Comment (NingLin-P) [reply frame/staking/src/tests.rs:5236]: Oh sorry, I thought this case was subject to the round down problem and changed it to meet the case `not affected chunk should not be slashed` (like https://github.com/paritytech/substrate/commit/bdd139989cb4dfa8cc6cc6b5ee470a711df32d25), should not changing it and add a new case instead.

File: frame/nomination-pools/test-staking/src/lib.rs [hunk 5]
```diff
@@ -490,12 +490,10 @@ fn pool_slash_proportional() {
...
-				// This last pool got slashed only the leftover dust. Otherwise in principle, this
-				// chunk/pool should have not been affected.
-				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 127, balance: 19 },
-				// This pool got slashed 12.5, which rounded down to 12.
-				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 128, balance: 8 },
-				// This pool got slashed 12.5, which rounded down to 12.
+				// This pool got slashed 12.5, which rounded up to 13.
+				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 128, balance: 7 },
+				// This pool got slashed 12 instead of 12.5 because earlier chunk got 0.5 more
+				// slashed, and 12 is all the remaining slash
```
Comment (kianenigma) [inline frame/nomination-pools/test-staking/src/lib.rs:496]:
```suggestion
				// This era got slashed 12.5, which rounded up to 13.
				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 128, balance: 7 },
				// This ear got slashed 12 instead of 12.5 because earlier chunk got 0.5 more
				// slashed, and 12 is all the remaining slash
```

File: frame/staking/src/tests.rs [hunk 6]
```diff
@@ -5233,18 +5246,19 @@ fn proportional_ledger_slash_works() {
 	ledger.active = unit;
 	ledger.total = unit * 4 + value;
 	// When
-	assert_eq!(ledger.slash(slash, 0, 0), slash - 43);
+	assert_eq!(ledger.slash(slash, 0, 0), slash - 5);
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5249]: why is this changed from 43 to 5?

**Review by kianenigma — APPROVED** [review]
LGTM, except the small question about tests.

File: frame/staking/src/tests.rs (hunk 6 above)
Comment (NingLin-P) [reply frame/staking/src/tests.rs:5249]: With rounding up each chunk will be slashed a bit more than before thus improving the accuration, the total slash will never exceed the expected amount though due to:
https://github.com/paritytech/substrate/blob/4c83ee0a406939f1393d19f87675e1fbc49e328d/frame/staking/src/lib.rs#L630

File: frame/nomination-pools/test-staking/src/lib.rs [hunk 7]
```diff
@@ -490,12 +490,10 @@ fn pool_slash_proportional() {
...
 			vec![
-				// This last pool got slashed only the leftover dust. Otherwise in principle, this
-				// chunk/pool should have not been affected.
-				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 127, balance: 19 },
-				// This pool got slashed 12.5, which rounded down to 12.
-				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 128, balance: 8 },
-				// This pool got slashed 12.5, which rounded down to 12.
+				// This era got slashed 12.5, which rounded up to 13.
+				PoolsEvent::UnbondingPoolSlashed { pool_id: 1, era: 128, balance: 7 },
+				// This ear got slashed 12 instead of 12.5 because earlier chunk got 0.5 more
```
Comment (ruseinov) [inline frame/nomination-pools/test-staking/src/lib.rs:495]:
```suggestion
				// This era got slashed 12 instead of 12.5 because an earlier chunk got 0.5 more
```

**Review by ruseinov — APPROVED** [review]
seems good!

**Comment by kianenigma** [comment]
bot rebase

**Comment by kianenigma** [comment]
bot merge
